cpp/src/arrow/util/functional.h

```diff
@@ -64,17 +71,47 @@ struct call_traits {
   /// extracted via call_traits::argument_type<Index, F>
   template <std::size_t I, typename F>
   using argument_type = decltype(argument_type_impl<I>(&std::decay<F>::type::operator()));
+
+  template <typename F>
+  using return_type = decltype(return_type_impl(&std::decay<F>::type::operator()));
 };
 
-template <typename T>
-struct type_constant {
-  using type = T;
+/// A callable object which forwards to a pointer and a member function pointer.
```

Comment:
Hmm, ok. `std::bind` can't work for this?

Reply:
I had honestly forgotten about `std::bind`. I will delete this class with glee

Reply:
Unfortunately, `std::bind` results in an overloaded function so its arguments can't be introspected with call_traits. I'm still deleting this class since I don't currently need to introspect any member functions except `operator()`.